crates/aptos-rest-client/src/lib.rs

```diff
@@ -69,46 +73,65 @@ pub struct Client {
     version_path_base: String,
 }
 
-impl Client {
-    pub fn new_with_timeout(base_url: Url, timeout: Duration) -> Self {
-        Client::new_with_timeout_and_user_agent(base_url, timeout, USER_AGENT)
-    }
-
-    pub fn new_with_timeout_and_user_agent(
-        base_url: Url,
-        timeout: Duration,
-        user_agent: &str,
-    ) -> Self {
-        let inner = ReqwestClient::builder()
-            .timeout(timeout)
-            .user_agent(user_agent)
-            .cookie_store(true)
-            .build()
-            .unwrap();
-
-        // If the user provided no version in the path, use the default. If the
-        // provided version has no trailing slash, add it, otherwise url.join
-        // will ignore the version path base.
-        let version_path_base = match base_url.path() {
-            "/" => DEFAULT_VERSION_PATH_BASE.to_string(),
-            path => {
-                if !path.ends_with('/') {
-                    format!("{}/", path)
-                } else {
-                    path.to_string()
-                }
-            },
-        };
+pub struct ClientBuilder {
+    reqwestBuilder: ReqwestClientBuilder,
+    version_path_base: String,
+    base_url: Url, // Default to mainnet if None
+    timeout: Option<Duration>, // Default to 10 seconds if None
+    headers: HeaderMap,
+}
+
+impl ClientBuilder {
+    pub fn new() -> Self {
```

Comment:
So with the builder pattern it is important to distinguish between:
1. Arguments that are required + have no sensible default.
1. Arguments that are optional + have no sensible default.
1. Arguments that are optional + have a sensible default.

Looking at all these arguments below:
- URL is 1
   - It is not possible to build a client without the user providing a URL. This means the builder should take the argument in to `new`.
- Timeout is 3
   - In `new` it makes sense to set a default, but we also allow people to override it.

As for `header` I think this is a helpful function (which would be 2), but we should also have a function on the builder for `X_APTOS_CLIENT`, like `x_aptos_client` (which would be 3). This way the caller doesn't have to know what header to use, that kinda thing. I'm less opinionated about this part though :P